### DISCUSSION TITLE
How can I get the Charachter in a Cells ?

### DISCUSSION THREAD

**Discussion by paule32** [magiblot/tvision#149] [Q&A] [answered]
Hello,
how can I get the Character under the Cursor based on cursor.x,, and cursor.y ?

**Comment by magiblot** [comment]
Hi Jens,

Are you referring to `TEditor` or to any view in general?

**Reply by paule32** [reply]
TEditr - TFileEditor

**Reply by magiblot** [reply]
Then you can get the caracter under the cursor with `bufChars(curPtr)`. This method returns a `TStringView` because some characters might be multi-byte (e.g. `€`, which is 3 bytes). However, be aware that this `TStringView` points to an internal buffer whose value will change on the next call to `bufChars`.

If you know that the character will only be one byte long, then you can also use `bufChar(burPtr)`, which returns a `char`.

**Reply by paule32** [reply]
and how can I set the Attributes of the Character(s) with ::setCell(Cells[selStart], 'A', Color); ?
where is Cells coming from ?

**Comment by magiblot** [comment]
The cells are stored in the `buffer` field. But changing the buffer manually won't give you the results you expect, because these changes will go away each time the view gets redrawn. Given that you modified `TEditor` in order to be able to override `formatLine()`, all styling should be performed in `formatLine`.

**Comment by magiblot** [comment]
What exactly are you trying to achieve? I believe you are going too far in adding hacks to `TEditor`. If I can understand what you need, I may be able to write an example of how to do it using https://github.com/magiblot/turbo.

**Comment by paule32** [comment]
I would implement a "select Text" Feature for the self made TSyntaxEditor.
In the EventHandler, I check for kbShift + Cursor left, and Cursor right.
Depend on this, I move the Cursor with setCursor.
Now, the Idea was, to change, and draw the Cells, instead using formatLine for text-marking.
```
void TSyntaxFileEditor::handleEvent( TEvent &event )
{
    TColorAttr  Color = 0x0f;

    if (event.what == evKeyDown) {
        if (event.keyDown.keyCode == kbLeft) {
            if ((event.keyDown.controlKeyState == 0x310)
            ||  (event.keyDown.controlKeyState == 0x330)
            ||  (event.keyDown.controlKeyState == 0x390)
            ||  (event.keyDown.controlKeyState == 0x110)
            ||  (event.keyDown.controlKeyState == 0x130)
            ||  (event.keyDown.controlKeyState == 0x190)) {
                clearEvent(event);
                
                selecting = true;
                selected  = true;

                selStart  = cursor.x - 1;
                selEnd    = cursor.x    ;
                
                setCursor(cursor.x-1,cursor.y);
            }
        }   else
        if (event.keyDown.keyCode == kbRight) {
            if ((event.keyDown.controlKeyState == 0x310)
            ||  (event.keyDown.controlKeyState == 0x330)
            ||  (event.keyDown.controlKeyState == 0x390)
            ||  (event.keyDown.controlKeyState == 0x110)
            ||  (event.keyDown.controlKeyState == 0x130)
            ||  (event.keyDown.controlKeyState == 0x190)) {
                clearEvent(event);

                TStringView chars = bufChars(curPtr);
                
                selecting = true;
                selected  = true;

                selStart  = cursor.x    ;
                selEnd    = cursor.x + 1;
                
                setCursor(cursor.x+1,cursor.y);
                
                ::setCell(Cells[selStart], 'A', Color);
            }
        }   else
        if (event.keyDown.keyCode == kbCtrlS) {
            clearEvent(event);
            messageBox("ctrl + s",mfInformation|mfOKButton);
            return;
        }
    }

    TFileEditor::handleEvent( event );
}
```

**Reply by paule32** [reply]
Hello,
a little step wider ...
I can "draw" a character at the cursor position, but not in the Color I would like it.
The color is red on white.
But it have the Attributes 0x00ff00, 0x000000 .
```
void TSyntaxFileEditor::handleEvent( TEvent &event )
{
    if (event.what == evKeyDown) {
        if (event.keyDown.keyCode == kbLeft) {
            if ((event.keyDown.controlKeyState == 0x310)
            ||  (event.keyDown.controlKeyState == 0x330)
            ||  (event.keyDown.controlKeyState == 0x390)
            ||  (event.keyDown.controlKeyState == 0x110)
            ||  (event.keyDown.controlKeyState == 0x130)
            ||  (event.keyDown.controlKeyState == 0x190)) {
                clearEvent(event);
                
                selecting = true;
                selected  = true;

                selStart  = cursor.x - 1;
                selEnd    = cursor.x    ;
                
                setCursor(cursor.x-1,cursor.y);
            }
        }   else
        if (event.keyDown.keyCode == kbRight) {
            if ((event.keyDown.controlKeyState == 0x310)
            ||  (event.keyDown.controlKeyState == 0x330)
            ||  (event.keyDown.controlKeyState == 0x390)
            ||  (event.keyDown.controlKeyState == 0x110)
            ||  (event.keyDown.controlKeyState == 0x130)
            ||  (event.keyDown.controlKeyState == 0x190)) {
                clearEvent(event);

                TStringView chars = bufChars(curPtr);
                
                selecting = true;
                selected  = true;

                selStart  = cursor.x    ;
                selEnd    = cursor.x + 1;
                
                setCursor(cursor.x+1,cursor.y);
 
                TColorAttr Color = { 0x00ff00, 0x000000 };
                writeChar(cursor.x, cursor.y, 'A', Color, 1);  // red white - but why ?
            }
        }   else
        if (event.keyDown.keyCode == kbCtrlS) {
            clearEvent(event);
            messageBox("ctrl + s",mfInformation|mfOKButton);
            return;
        }
    }

    TFileEditor::handleEvent( event );
}
```
The following is a Demo-Snipp:
![grafik](https://github.com/magiblot/tvision/assets/3682603/c81dbe4f-f1a7-467d-a35f-b605d1a204ca)

**Reply by magiblot** [reply, marked as answer]
For the record, the reason why that color is not displayed as expected is explained here: https://github.com/magiblot/tvision/discussions/150#discussioncomment-8357098

**Comment by magiblot** [comment]
Hi Jens!

Sorry for the late reply.

As I said, I have written an example of a simple code editor using https://github.com/magiblot/turbo to see if it may be of help to you.

Using that repository's latest commit, if you build it with `-DTURBO_BUILD_EXAMPLES=ON`:
```bash
cmake . -B build -DTURBO_BUILD_EXAMPLES=ON
cmake --build build
```

In the build directory you should find a program named `demo-pascal-editor`. The source code of that program is located in https://github.com/magiblot/turbo/blob/master/source/examples/demo-pascal-editor.cc.

I suggest you check it out, and if that program does more or less the same you are trying to achieve, then it might be easier for you to reuse the Turbo editor component than to hack `TEditor`.

Cheers.
